email_alert_service/models/document_links.rb

```diff
@@ -0,0 +1,53 @@
+class DocumentLinks
+  attr_reader :document
+
+  def initialize(document)
+    @document = document
+  end
+
+  def self.call(args)
+    new(args).links
+  end
+
+  def links
+    populate_links.merge("taxon_tree" => taxon_tree)
+  end
+
+private
+
+  def populate_links
+    raw_links.merge(reverse_links_hash)
```

Comment:
I'd wonder if you can simplify this by not needing to use the raw_links anymore and instead get all the links from expanded_links.

So I guess loop through all expanded_links and get content_ids, and remove available_translations. 

I think that might allow the code to be simpler.

Reply:
As per your note I've amended, so that we are now building the links payload from an empty hash, into which we add content-ids taken from the expanded links. Rather than starting with a hash of links from publishing API, and then adding extra expanded links into that hash.

However I noticed on the screen shots of `ContentChanges` added to the PR description, that there's a difference in the contents of the `taxon` key.

Main
```
"taxons"=>
    ["3dbeb4a3-33c0-4bda-bd21-b721b0f8736f",
     "d956c72a-246d-4787-af39-00bf58b2ea66",
     "7ae55a0e-e4d2-49ab-8364-c0aa307bf70b",
     "34d0a5ae-c166-4830-be69-695971b46666",
     "06ad07f7-1e79-462f-a192-6b2c9d92089c"],
   "topics"=>["0e392f84-8f09-49f7-82f4-ce6765e23d3e"],
   "taxon_tree"=>
    ["3dbeb4a3-33c0-4bda-bd21-b721b0f8736f",
     "9597c30a-605a-4e36-8bc1-47e5cdae41b3",
     "d956c72a-246d-4787-af39-00bf58b2ea66",
     "7ae55a0e-e4d2-49ab-8364-c0aa307bf70b",
     "05d212cd-d06e-43b6-a31a-c3bfc2ab1d3c",
     "91b8ef20-74e7-4552-880c-50e6d73c2ff9",
     "34d0a5ae-c166-4830-be69-695971b46666",
     "67b24f42-ff20-4ce6-be93-4f505b388ab5",
     "ce9e9802-6138-4fe9-9f33-045ef213be29",
     "6e3f3cfb-142a-41f4-a03a-fea504cc1f79"],
```
Branch
```
"taxons"=>
    ["3dbeb4a3-33c0-4bda-bd21-b721b0f8736f",
     "d956c72a-246d-4787-af39-00bf58b2ea66",
     "7ae55a0e-e4d2-49ab-8364-c0aa307bf70b",
     "34d0a5ae-c166-4830-be69-695971b46666"],
   "topics"=>["0e392f84-8f09-49f7-82f4-ce6765e23d3e"],
   "taxon_tree"=>
    ["3dbeb4a3-33c0-4bda-bd21-b721b0f8736f",
     "9597c30a-605a-4e36-8bc1-47e5cdae41b3",
     "d956c72a-246d-4787-af39-00bf58b2ea66",
     "7ae55a0e-e4d2-49ab-8364-c0aa307bf70b",
     "05d212cd-d06e-43b6-a31a-c3bfc2ab1d3c",
     "91b8ef20-74e7-4552-880c-50e6d73c2ff9",
     "34d0a5ae-c166-4830-be69-695971b46666",
     "67b24f42-ff20-4ce6-be93-4f505b388ab5",
     "ce9e9802-6138-4fe9-9f33-045ef213be29",
     "6e3f3cfb-142a-41f4-a03a-fea504cc1f79"],
```

ie `"06ad07f7-1e79-462f-a192-6b2c9d92089c"` is missing from the list of taxons for the branch code. This makes sense when you [look what was sent from Publishing API to Email alert service](https://deploy.integration.publishing.service.gov.uk/job/run-rake-task/305189/console), as that taxon is not present in the expanded links. It is only present in the links, which we are now overwriting.

I am unclear how there can be a discrepancy between the links and the expanded links. Ie I'm not sure if this is an "integration is sometimes a bit weird and out of sync" problem, or if I'm fundamentally misunderstanding the difference between links and expanded links. Do you know @kevindew ?

For ref, this is the response from publishing_api when I run a `get_content` for the "missing" taxon:

```
irb(main):007:0> Services.publishing_api.get_content("06ad07f7-1e79-462f-a192-6b2c9d92089c")
{"request_uri":"https://publishing-api.integration.govuk-internal.digital/v2/content/06ad07f7-1e79-462f-a192-6b2c9d92089c","start_time":1669732555.4125526,"status":404,"end_time":1669732555.464463,"body":"{\"error\":{\"code\":404,\"message\":\"Could not find locale: en for document with content_id: 06ad07f7-1e79-462f-a192-6b2c9d92089c\"}}"}
/data/vhost/collections-publisher/shared/bundle/ruby/3.1.0/gems/gds-api-adapters-84.0.0/lib/gds_api/json_client.rb:108:in `rescue in do_json_request': URL: https://publishing-api.integration.govuk-internal.digital/v2/content/06ad07f7-1e79-462f-a192-6b2c9d92089c (GdsApi::HTTPNotFound)
Response body:
{"error":{"code":404,"message":"Could not find locale: en for document with content_id: 06ad07f7-1e79-462f-a192-6b2c9d92089c"}}
```

Reply:
Excellent thanks @kevindew for explaining